src/contracts/GPv2Settlement.sol

```diff
@@ -0,0 +1,91 @@
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@uniswap/v2-core/contracts/interfaces/IUniswapV2Factory.sol";
+
+/// @title Gnosis Protocol v2 Settlement Contract
+/// @author Gnosis Developers
+contract GPv2Settlement {
+    /// @dev The domain separator used for signing orders that gets mixed in
+    /// making signatures for different domains incompatible.
+    bytes32 public constant DOMAIN_SEPARATOR = keccak256("GPv2");
+
+    /// @dev The stride of an encoded order.
+    uint256 private constant ORDER_STRIDE = 120;
+
+    /// @dev The Uniswap factory. This is used as the AMM that GPv2 settles with
+    /// and is responsible for determining the range of the settlement price as
+    /// well as trading surplus that cannot be directly settled in a batch.
+    IUniswapV2Factory public immutable uniswapFactory;
+
+    /// @dev The global nonce for determining the current batch. This is used to
+    /// ensure orders can't be replayed in multiple batches.
+    uint256 public nonce;
+
+    /// @param uniswapFactory_ The Uniswap factory to act as the AMM for this
+    /// GPv2 settlement contract.
+    constructor(IUniswapV2Factory uniswapFactory_) public {
+        uniswapFactory = uniswapFactory_;
+    }
+
+    /// @dev Settle the specified orders at a clearing price. Note that it is
+    /// the responsibility of the caller to ensure that all GPv2 invariants are
+    /// upheld for the input settlement, otherwise this call will revert.
+    /// Namely:
+    /// - The prices are better than, or equal to, the AMM's price with its fee
+    ///   spread. Specifically the price is either:
+    ///   - Effective price in case the AMM was traded with,
+    ///   - Spot price in case the AMM was not traded with.
+    /// - All orders are valid and signed
+    /// - Accounts have sufficient balance and approval.
+    ///
+    /// Note that orders in this method are expressed as encoded bytes. These
+    /// bytes are tightly packed orders encoded with `abi.encodePacked` in order
+    /// to reduce the amount of call data required to call this method. Orders
+    /// encode the following fields:
+    /// ```
+    /// struct Order {
+    ///     sellAmount: uint112,
+    ///     buyAmount:  uint112,
+    ///     validTo:    uint32,
+    ///     nonce:      uint32,
+    ///     tip:        uint112,
+    ///     validTo:    uint32,
+    ///     flags:      uint8,
+    ///     signature: {
+    ///         v: uint8,
+    ///         r: bytes32,
+    ///         s: bytes32,
+    ///     }
+    /// }
+    /// ```
+    ///
+    /// Note that the encoded order data does not contain which token is the
+    /// sell or buy token. This data is implicit from which parameter the order
+    /// was specified from.
+    ///
+    /// @param token0 The address of token 0 being traded in the batch.
+    /// @param token1 The address of token 1 being traded in the batch.
+    /// @param d0 The amount of token 0 being traded to the AMM.
```

Comment:
We could define that the d0>= 0 and d1 =0 for all cases, and hence leave d1 out of the spec?

Reply:
I thought about that, but wasn't sure how Uniswap pairs are computed and if `USDT-WETH` and `WETH-USDT` are equivalent.

Also, the small advantage of using d0 and d1 is that we are computing them anyway when finding the price, so we can reuse this computation and save a bit of gas.

Reply:
>Also, the small advantage of using d0 and d1 is that we are computing them anyway when finding the price, so we can reuse this computation and save a bit of gas.

Oh right, uniswap is not calculating these numbers for us, we need to know them anyways. Cool, then let's keep it

Reply:
To me this implies that Uniswap pairs do have an order, since for any tokens A and B, there only exists one token pair (with the address being computed based on the addresses of A and B **in lexicographical order**), since the UniswapV2Pair contract has a token chosen as `token0` and the other as `token1`.

Reply:
I'd argue that since given two tokens there only exists one token pair, then the order of the two is not really important. Conversely, if they weren't sorted then the order would matter since the pair A-B and the pair B-A would have a different contract.
I don't think that the order of the tokens currently matters for anything else but computing the salt.

Reply:
> Conversely, if they weren't sorted then the order would matter since the pair A-B and the pair B-A would have a different contract.

Hehe, I think we are arguing two sides to the same coin. My point was that when we interact with the `IUniswapV2Pair` the order of `token0` and `token1` matters, as the pair's `reserve0` and `reserve1` is fixed for a token pair. That is, for some tokens A and B:
- There is one Unsiwap pair contract with A = tokenX and B = tokenY, with X != Y and X, Y ∈ {0, 1}
- The contract for the pair A-B == B-A
- Since the pair's `reserveX` is in token A and `reserveY` is in token B, then `d0` and `d1` parameters need to either match the Uniswap pair's `reserve0` and `reserve1` (my preference) or we need to check if `calldata.token0 == pair.token0`, and if not, inverse the input tokens (so 0 -> 1 and 1 -> 0).

Does that make sense?

Reply:
Very true, we most likely need to add some logic from our side to tell which is token0 and whick is token1.